langchain/retrievers/google_cloud_enterprise_search.py

```diff
@@ -0,0 +1,188 @@
+"""Retriever wrapper for Google Cloud Enterprise Search on Gen App Builder."""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Dict, List, Sequence
+
+from pydantic import Extra, root_validator, Field
+
+from langchain.callbacks.manager import (
+    AsyncCallbackManagerForRetrieverRun,
+    CallbackManagerForRetrieverRun,
+)
+from langchain.schema import BaseRetriever, Document
+from langchain.utils import get_from_dict_or_env
+
+if TYPE_CHECKING:
+    from google.cloud.discoveryengine_v1beta import (
+        SearchServiceClient,
+        SearchRequest,
+        SearchResult,
+    )
+
+
+class GoogleCloudEnterpriseSearchRetriever(BaseRetriever):
+    """Wrapper around Google Cloud Enterprise Search Service API.
+    For the detailed explanation of the Enterprise Search concepts
+    and configuration parameters refer to the product documentation.
+
+    https://cloud.google.com/generative-ai-app-builder/docs/enterprise-search-introduction
+    """
+
+    project_id: str
+    """Google Cloud Project ID."""
+    search_engine_id: str
+    """Enterprise Search engine ID."""
+    serving_config_id: str = "default_config"
+    """Enterprise Search serving config ID."""
+    location_id: str = "global"
+    """Enterprise Search engine location."""
+    filter: str = None
+    """Filter expression."""
+    get_extractive_answers: bool = False
+    """If True the retriever will return Extractive Answers. Otherwise Extractive Segments."""
+    max_documents: int = Field(default=5, ge=1, le=100)
```

Comment:
most retriever / doc store interfaces refer to this as `top_k`, could we change for consistency?

and out of curiosity, are the bounds enforced on the server side as well? if not, do we need them?

Reply:
@baskaryan 

This setting is a little bit different than top_k  as the underlying retrieval mechanism is a little bit different than in many other vector stores . The max_documents, max_extractive_answer_count, and max_extractive_segment_count settings together decide how many matching "chunks" will be returned. The max_documents controls how many top ranked documents - as ingested by Enterprise Search engine - will be used to prepare extractive answers or extractive segments. When you populate Enterprise Search you ingest full document (rather than chunks) and the engine does chunking and embedding generations in the backend. 

For example, if you set max_documents to 3 and max_extractive_answers_count to 3 you can get up to 9 LangChain documents.

When we have the retriever in the main, we are planning to publish a number of code samples that show how to use different LangChain retrieval chains and agents with Enterprise Search. In these code samples we plan to go into pretty deep detail in describing different strategies for ingesting/quering both structured and unstructured data with Enterprise Search.